### PR TITLE
Structured event stream for judges [Spec + Impl]

### PR DESCRIPTION
**Pull request by rhuss**
## Summary

Start with the [Review Guide (REVIEWERS.md)](https://github.com/rhuss/agent-eval-harness/blob/002-structured-events/specs/002-structured-events/REVIEWERS.md) for a quick overview of what this feature does, key decisions, and areas needing attention.

- Add specification and design documents for the `traces.events` feature: parse stream-json JSONL at collection time into structured `events.json` per case
- Consolidates 4 redundant JSONL parsers (`extract_usage`, `_extract_tool_calls`, `_extract_assistant_text`, `build_trace`) into a shared `agent_eval/events.py` module
- Events use a flat-with-tags structure: root and subagent events in one chronological list, subagent events tagged with `parent_tool_use_id` and `agent_id`
- Introduces `{{ conversation }}` template variable (replaces `{{ stdout }}`), removes `record["stdout"]` from judges
- Brainstorm doc for four new judge patterns enabled by structured events (process quality, cost/efficiency, safety/compliance, regression fingerprinting)

**Context:** This was identified during the `{{ stdout }}` implementation (PR #57). Antonin's original vision was to parse stream-json at collection time so judges don't have to do the parsing individually. We discussed the subagent event structure and agreed on flat-with-tags as the simplest approach that matches Claude Code's native streaming format.

**This PR contains spec/design documents only, no implementation code.** The spec includes 29 tasks across 4 user stories, ready for implementation.

## Artifacts

- [`REVIEWERS.md`](https://github.com/rhuss/agent-eval-harness/blob/002-structured-events/specs/002-structured-events/REVIEWERS.md) - Start here: review guide with key decisions and areas needing attention
- `specs/002-structured-events/spec.md` - Feature specification (4 user stories, 16 FRs)
- `specs/002-structured-events/plan.md` - Implementation plan
- `specs/002-structured-events/tasks.md` - 29 tasks in 8 phases
- `specs/002-structured-events/data-model.md` - Event schema with examples
- `specs/002-structured-events/research.md` - 9 design decisions with rationale
- `brainstorm/02-event-powered-judges.md` - Future judge patterns with example code

## Test plan

- [ ] Spec reviewed for completeness and implementability
- [ ] Event schema validated against actual Claude Code stream-json output
- [ ] Subagent flat-with-tags approach confirmed with Antonin

Assisted-By: 🤖 Claude Code

## Summary by CodeRabbit

* **Documentation**
  * Added brainstorming, design, spec, data-model, reviewer guide, checklist, task plan, and research docs for a collection-time structured event stream for judges.
  * Documented judge-facing changes: new conversation template variable, event-based outputs, event/result size cap with truncation/placeholder behavior, flat root+subagent event ordering with deduplication, parsing/loading/edge-case rules, and reviewer checklist and migration guidance.

### CODE REVIEW

File: brainstorm/01-structured-events.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+**Approach C: Shared parser module + collection-time storage.**
+
+The redundancy of 4 independent JSONL parsers is the root problem, and only a shared parser addresses it structurally. The extra scope is manageable because the parsing logic already exists across the codebase (consolidation, not invention).
+
+### Key Design Decisions
+
+**Events replace stdout for judges.** `record["stdout"]` is removed from the record dict. Judges access execution data exclusively through `record["events"]`. If a judge needs data that events don't expose, we extend the events schema rather than fall back to raw JSONL. No escape hatch.
+
+**`{{ stdout }}` is deprecated and errors.** The `{{ stdout }}` template variable (added in PR #57) was a workaround for the absence of structured events. Once events land, `{{ stdout }}` should raise an error pointing judge authors to the events-based replacement (e.g., a new `{{ conversation }}` template variable or direct `outputs["events"]` access in check judges). This avoids quietly reimplementing a raw-stdout workaround on top of the structured interface.
```
[the remark this answers is not in the document]
Comment (rhuss) [reply brainstorm/01-structured-events.md:60]: Intentional design decision. We discussed this explicitly during brainstorming and chose the hard break over a deprecation grace period. The reasoning: a soft deprecation lets judges silently keep parsing raw JSONL, which undermines the structured events interface. If events are missing data, the correct fix is extending the events schema, not falling back to raw parsing. The error message on KeyError will include clear migration guidance pointing to `outputs["events"]` and `{{ conversation }}`.

File: brainstorm/01-structured-events.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+**Raw stdout stays on disk for debugging.** `traces.stdout: true` controls whether `stdout.log` is kept on disk for human inspection. `traces.events: true` controls whether structured events are parsed and available to judges. Both can be true independently.
+
+**Tool results included with a size cap.** Tool result content is stored in events with a generous cap (e.g., 50K chars per result). This covers all realistic judge scenarios without bloating `events.json` from pathological cases (e.g., reading a 500K file).
+
+**Subagent events included as flat-with-tags.** All events (root + subagent) in one ordered list, subagent events tagged with `parent_tool_use_id` and `agent_id`. Subagent transcript files merged and deduplicated by message ID. This matches Claude Code's native streaming format and keeps judge iteration simple.
```
[the remark this answers is not in the document]
Comment (rhuss) [reply brainstorm/01-structured-events.md:66]: The threat model for this tool doesn't include adversarial runners. The JSONL comes from Claude Code running inside the eval harness, not from untrusted external sources. A compromised runner has far more direct attack vectors than crafting tool output to exploit truncation. That said, we added truncation metadata (`truncated: true`, `original_length`) in the latest commit so judges can detect when content was capped and act accordingly.

File: brainstorm/02-event-powered-judges.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    if e["type"] != "assistant":
+        continue
+    for tool in e.get("tools", []):
+        if tool["name"] == "Bash":
+            cmd = tool.get("input", {}).get("command", "")
+            for pattern in dangerous_patterns:
+                if pattern in cmd:
+                    return (False, f"Dangerous command detected: {cmd[:100]}")
+return (True, "No dangerous commands found")
+```
```
[the remark this answers is not in the document]
Comment (rhuss) [reply brainstorm/02-event-powered-judges.md:89]: Good point, but this is a brainstorm doc showing illustrative examples of what event-powered judges could look like, not production code. The example intentionally uses a simple pattern to show the concept. A real safety judge would need more sophisticated analysis. We'll address that when implementing the actual judge patterns.

File: specs/002-structured-events/data-model.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+```json
+{
+  "type": "assistant",
+  "text": "extracted assistant text (concatenated from text blocks)",
+  "tools": [
+    {
+      "name": "Read",
+      "id": "toolu_abc123",
+      "input": {"file_path": "/path/to/file"}
+    }
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/data-model.md:19]: Events are read-only data for judges. Judges receive the event list as a Python dict and use it for evaluation logic (counting tool calls, checking patterns, reading text). They don't execute tool inputs or use file paths to access the filesystem. The file paths in tool inputs are historical data showing what the skill did during execution, not instructions for the judge to follow.

File: specs/002-structured-events/data-model.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  stderr: true              # Keep raw stderr.log on disk
+  events: true              # Parse JSONL into events.json (default: true)
+  event_result_cap: 50000   # Max chars per tool result content
+  metrics: true             # Capture run_result.json metrics
+```
+
+## Relationships
+
+- Each **tool_result** event references a **tool call** in a preceding **assistant** event via `tool_use_id`
+- The **events list** for a case combines the case's `stdout.log` content with events from `subagents/*.jsonl` transcript files (deduplicated by message ID)
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/data-model.md:118]: Message IDs come from the Claude API (`msg_` prefix, globally unique). They're not generated by the harness or by subagents. The deduplication pattern (`seen_msg_ids`) is already proven in `stream_capture.py` and `count_subagent_turns()` across the existing codebase without collision issues.

File: specs/002-structured-events/plan.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+Add a shared JSONL parser (`agent_eval/events.py`) that converts Claude Code stream-json stdout into structured events at collection time. Events are stored as `events.json` per case and loaded into `record["events"]` for judges. The event list uses a flat-with-tags structure: root and subagent events in one ordered list, subagent events tagged with `parent_tool_use_id` and `agent_id`. Subagent transcript files are merged and deduplicated.
+
+This replaces `record["stdout"]` (removed, KeyError on access), consolidates `_extract_tool_calls()` and `_extract_assistant_text()` into event lookups, and introduces `{{ conversation }}` as the template variable for LLM judges (deprecating `{{ stdout }}` with an error).
+
+## Technical Context
+
+**Language/Version**: Python 3.12+
+**Primary Dependencies**: json (stdlib), yaml
+**Storage**: File-based (`events.json` per case in run output directory)
+**Testing**: pytest (unit tests in `tests/`, e2e tests in `tests/e2e/`)
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/plan.md:17]: Valid concern for production deployments, but out of scope for this feature. The eval harness runs locally (or in CI) against test cases, not in shared environments with untrusted access to run artifacts. The existing `stdout.log`, `run_result.json`, and collected artifacts already persist potentially sensitive content with the same file permissions. If artifact hardening becomes a requirement, it should be a cross-cutting concern applied to all run artifacts, not just `events.json`.

File: specs/002-structured-events/research.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+**Decision**: `{{ conversation }}` replaces `{{ stdout }}`. Using `{{ stdout }}` raises an error with migration guidance.
+
+**Rationale**: "conversation" clearly describes what it renders (assistant conversation text). It's distinct from `{{ events }}` (which could be confused with the full event list) and from `{{ transcript }}` (which could be confused with subagent transcripts).
+
+## R7: stdout removal from record
+
+**Decision**: `record["stdout"]` is removed entirely. Accessing `outputs["stdout"]` in a judge raises KeyError. No deprecation grace period.
+
+**Rationale**: A hard break forces immediate migration and prevents the raw JSONL escape hatch from undermining the structured events interface. If events are missing data, the correct fix is extending the events schema, not falling back to raw parsing.
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/research.md:55]: Same reasoning as the brainstorm comment above. This is an intentional design decision, not an oversight. The hard break forces migration to the structured interface rather than letting raw JSONL parsing persist as a shadow API.

File: specs/002-structured-events/research.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+**Alternatives considered**:
+- Nested (subagent events inside Agent tool result): forces recursion for simple queries like "count all tool calls"
+- Separate lists by agent ID: loses chronological ordering between root and subagent activity
+
+## R9: Subagent transcript deduplication
+
+**Decision**: Reuse the existing `seen_msg_ids` pattern from `stream_capture.py`. When merging transcript files, skip events whose message ID was already seen in the stdout stream.
+
+**Rationale**: Claude Code >= 2.1.108 streams foreground subagent messages in stdout with `parent_tool_use_id`. The same messages appear in `subagents/*.jsonl` transcripts. Deduplication by message ID is already proven in `extract_usage()` and `count_subagent_turns()`.
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/research.md:71]: Message IDs from Claude's API are globally unique (`msg_` prefixed UUIDs). The `seen_msg_ids` deduplication pattern has been working reliably in `stream_capture.py` since the harness was created. Adding composite keys (message_id + agent_id) would add complexity without solving a real problem, since ID collisions don't occur in practice.

File: specs/002-structured-events/tasks.md [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+(No setup tasks required.)
+
+---
+
+## Phase 2: Foundational (Shared Parser Module)
+
+**Purpose**: Create the shared event parser that all user stories depend on.
+
+- [ ] T001 Create `agent_eval/events.py` with `parse_stream_events(stdout_text, result_cap=50000)` function that parses JSONL into structured event dicts per the data model schema (includes subagent events with `parent_tool_use_id` and `agent_id` tags)
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/tasks.md]: The JSONL comes from Claude Code running inside the harness, not from untrusted external sources. A compromised runner could manipulate any part of the output, not just subagent tags. Validating `parent_tool_use_id` references against actual Agent tool calls would add complexity for a threat that doesn't exist in this execution model.

File: specs/002-structured-events/tasks.md [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+---
+
+## Phase 7: Edge Cases
+
+**Purpose**: Verify graceful handling of edge cases identified in the spec.
+
+- [ ] T020 [P] Unit test for empty/missing stdout.log in `tests/test_events.py`: verify `parse_stream_events("")` returns `[]`
+- [ ] T021 [P] Unit test for non-JSONL content in `tests/test_events.py`: verify plain text lines are skipped, returns `[]`
+- [ ] T022 [P] Unit test for tool result exceeding 50K cap in `tests/test_events.py`: verify content truncated with `"[truncated]"` marker
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/tasks.md]: Good point. We addressed this partially: malformed `events.json` now logs a warning to stderr (updated FR-003). For the distinction between "no events" vs "parse found nothing", the practical impact is low since both mean the same thing for judges (no conversation data to evaluate). If needed later, we could add metadata to events.json (e.g., `{"events": [], "parse_stats": {"lines": 0}}`) but that's premature for v1.

File: specs/002-structured-events/spec.md [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+- What is the difference between foreground and background subagents? Foreground subagents (Claude Code >= 2.1.108) are streamed in stdout with `parent_tool_use_id`. Background subagents only appear in `subagents/*.jsonl` transcript files. Both types get `parent_tool_use_id` and `agent_id` tags in the event list. The distinction is transparent to judges.
+
+## Requirements *(mandatory)*
+
+### Functional Requirements
+
+- **FR-001**: The system MUST provide a shared parsing function that converts JSONL stdout into a list of structured event objects
+- **FR-002**: The system MUST parse events at collection time and persist them as `events.json` per case in the run output directory using atomic writes (write to temp file, then rename)
+- **FR-003**: The scoring system MUST load `events.json` into `record["events"]` when building the case record. If `events.json` is missing or malformed, `record["events"]` MUST be set to `[]` and a warning logged to stderr
+- **FR-004**: The system MUST remove `record["stdout"]` from the case record. Accessing `outputs["stdout"]` in a judge MUST raise a KeyError. Judges access execution data exclusively through `record["events"]`
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/spec.md]: Not a real inconsistency. `record` and `outputs` refer to the same dict at different points in the pipeline. `load_case_record()` builds the `record` dict, then `score_cases()` passes it to judges as the `outputs` kwarg. So `record["events"]` is the internal name in score.py, `outputs["events"]` is what judges see. This is the existing pattern for all fields (`record["files"]` / `outputs["files"]`, `record["tool_calls"]` / `outputs["tool_calls"]`, etc.). The spec uses `record["events"]` when describing the data pipeline and `outputs["events"]` when describing the judge interface, which is correct.

File: brainstorm/01-structured-events.md [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+**`trace_builder.py` stays independent for now.** It builds MLflow-specific hierarchical spans, which is a different shape than the flat event list judges need. It could adopt the shared parser later as an optional refactor.
+
+## Key Requirements
+
+1. New `agent_eval/events.py` module with `parse_stream_events(stdout_text) -> list[Event]`
+2. Events parsed at collection time by `collect.py`, stored as `events.json` per case
+3. `load_case_record()` loads events into `record["events"]`, removes `record["stdout"]`
+4. `{{ stdout }}` template variable deprecated: raises an error with migration guidance
```
[the remark this answers is not in the document]
Comment (rhuss) [reply brainstorm/01-structured-events.md:75]: Addressed in commit 6f9e31a. FR-002 now specifies atomic writes (temp file + rename).

File: specs/002-structured-events/data-model.md [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+  "type": "tool_result",
+  "tool_use_id": "toolu_abc123",
+  "tool_name": "Read",
+  "content": "file contents here...",
+  "is_error": false,
+  "timestamp": "2026-05-06T10:00:01.000Z"
+}
+```
+
+- `content`: The tool's output, capped at 50K chars (default). If truncated, ends with `"[truncated]"`.
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/data-model.md]: Addressed in commit 6f9e31a. FR-009 now requires `truncated: true` and `original_length: N` metadata on capped results so judges can detect truncation. Non-UTF-8 content gets a binary placeholder. The threat model doesn't include adversarial runners (JSONL comes from Claude Code inside the harness).

File: specs/002-structured-events/REVIEWERS.md [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+- **Breaking change impact**: Any check judge parsing `outputs.get("stdout", "")` will break. Error messages must include clear migration guidance.
+
+- **Event schema as contract**: Once `events.json` is written by collect.py and consumed by score.py, the schema becomes a versioning concern. Older files from previous runs may not be compatible if the schema evolves.
+
+- **50K tool result cap**: Generous but arbitrary. Judges evaluating large file contents could lose data beyond the cap. The configurable `event_result_cap` mitigates this, but the default could surprise users.
+
+- **`traces.events` defaulting to `true`**: Old eval runs without `events.json` will have `record["events"]` as an empty list when re-scored. Judges should handle this gracefully.
+
+- **Subagent deduplication edge cases**: Claude Code >= 2.1.108 streams foreground subagent messages in stdout, but background agents only appear in transcript files. The deduplication must handle both cases correctly.
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/REVIEWERS.md]: Addressed in commit 6f9e31a. Added edge case clarification: foreground subagents (>= 2.1.108) stream in stdout with `parent_tool_use_id`, background agents only appear in transcript files. Both get `parent_tool_use_id` and `agent_id` tags. Transparent to judges.

File: specs/002-structured-events/REVIEWERS.md [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+- [ ] Breaking changes are documented with migration guidance
+- [ ] Event schema is stable enough to be a contract
+- [ ] Scope matches the stated boundaries
+- [ ] Success criteria are achievable
+- [ ] No unstated assumptions
+- [ ] Backward compatibility for `outputs["tool_calls"]` is maintained
+- [ ] Edge cases (empty stdout, non-JSONL, subagent filtering, deduplication) are tested
+- [ ] Subagent events correctly tagged with `parent_tool_use_id` and `agent_id`
+- [ ] `{{ conversation }}` renders root-only text (no subagent text)
+
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/REVIEWERS.md:92]: Addressed in commit 6f9e31a. Added checklist items for atomic writes, schema validation on load, non-UTF-8 handling, and truncation metadata.

File: specs/002-structured-events/spec.md [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+- **FR-002**: The system MUST parse events at collection time and persist them as `events.json` per case in the run output directory
+- **FR-003**: The scoring system MUST load `events.json` into `record["events"]` when building the case record
+- **FR-004**: The system MUST remove `record["stdout"]` from the case record. Accessing `outputs["stdout"]` in a judge MUST raise a KeyError. Judges access execution data exclusively through `record["events"]`
+- **FR-005**: The system MUST provide a `{{ conversation }}` template variable for LLM judge prompts that renders assistant conversation text from the structured events
+- **FR-006**: The `{{ stdout }}` template variable MUST raise an error with a migration message when used, guiding authors to use `{{ conversation }}`
+- **FR-007**: Each structured event MUST include: event type, content (text or tool data), and timestamp when available
+- **FR-008**: Tool call events MUST include the tool name and input parameters
+- **FR-009**: Tool result events MUST include the result content, capped at 50K characters by default (configurable via `traces.event_result_cap` in eval.yaml) with a `"[truncated]"` marker for oversized results
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/spec.md]: Addressed in commit 6f9e31a. FR-009 updated to require valid UTF-8 text, non-UTF-8 replaced with binary placeholder, truncation metadata fields added. Tool inputs (FR-008) are stored as-is from Claude's API output since events are read-only data for judges, not executed. No path traversal risk.

File: specs/002-structured-events/spec.md [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+
+## Success Criteria *(mandatory)*
+
+### Measurable Outcomes
+
+- **SC-001**: Judges can access structured events via `outputs["events"]` without writing any JSONL parsing code
+- **SC-002**: JSONL parsing occurs exactly once per case (at collection time), not at scoring time
+- **SC-003**: Existing check judges that used `outputs["tool_calls"]` continue to work (tool calls derived from events)
+- **SC-004**: LLM judges using `{{ conversation }}` receive assistant conversation text identical to what `{{ stdout }}` previously rendered
+- **SC-005**: Collection time overhead for event parsing adds no more than 500ms per case
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/spec.md]: Addressed in commit 6f9e31a. SC-005 updated to: "Event parsing overhead scales linearly with stdout size, verified by benchmark test against representative JSONL sizes (1KB, 100KB, 1MB)." Added task T027 for the benchmark.

File: specs/002-structured-events/tasks.md [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+- [ ] T023 [P] Unit test for `traces.events: false` in `tests/test_events.py`: verify no `events.json` written, `record["events"]` is `[]`
+- [ ] T024 [P] Unit test for subagent deduplication in `tests/test_events.py`: verify events streamed in stdout and also in transcript are not double-counted
+
+---
```
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/tasks.md:130]: Addressed in commit 6f9e31a. Added T027 benchmark test for parse_stream_events() with representative JSONL sizes.

File: specs/002-structured-events/spec.md [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+
+- Q: What should the new template variable replacing `{{ stdout }}` be named? → A: `{{ conversation }}` (renders assistant text from events, distinct from the full `outputs["events"]` list)
+- Q: What is the default size cap for tool results in events? → A: 50K characters default, configurable via `traces.event_result_cap` in eval.yaml
+- Q: What happens when a check judge accesses `outputs["stdout"]`? → A: Raises KeyError immediately (hard break, forces migration)
+
+### Session 2026-05-08
+
+- Q: How should subagent events be structured? → A: Flat list with tags. All events (root + subagent) in one ordered list, subagent events tagged with `parent_tool_use_id` and `agent_id`. Judges filter to root-only with `if not e.get("parent_tool_use_id")`. Matches Claude Code's native streaming format (>= 2.1.108). Alternatives rejected: nesting (forces recursion for simple queries), separate lists by agent (loses ordering).
+
+## User Scenarios & Testing *(mandatory)*
```
Comment (astefanutti) [inline specs/002-structured-events/spec.md:25]: I wonder whether this is the moment to start trying to use agent-eval-harness to have e2e tests for agent-eval-harness?

File: specs/002-structured-events/spec.md [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+**Feature Branch**: `002-structured-events`
+**Created**: 2026-05-06
+**Status**: Draft
+**Input**: Brainstorm document `brainstorm/01-structured-events.md`
+
+## Clarifications
+
+### Session 2026-05-06
+
+- Q: What should the new template variable replacing `{{ stdout }}` be named? → A: `{{ conversation }}` (renders assistant text from events, distinct from the full `outputs["events"]` list)
```
Comment (astefanutti) [inline specs/002-structured-events/spec.md]: Just to make sure before we commit to `events`, would `outputs["turns"]` be conceptually and conventionally closer?

File: specs/002-structured-events/spec.md [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+A judge author uses `{{ conversation }}` in an LLM judge prompt to include the skill's conversation output. The template variable renders the assistant text from the structured events, replacing the deprecated `{{ stdout }}` approach.
+
+**Why this priority**: LLM judges need a template variable to include conversation text in their prompts. This replaces `{{ stdout }}` with an events-backed equivalent.
+
+**Independent Test**: Configure an LLM judge with the new template variable. Run scoring. Verify the rendered prompt contains the assistant conversation text and not raw JSONL.
+
+**Acceptance Scenarios**:
+
+1. **Given** an LLM judge prompt containing `{{ conversation }}`, **When** scoring runs, **Then** the variable is replaced with concatenated assistant text from the structured events
+2. **Given** a judge prompt containing `{{ stdout }}`, **When** scoring runs, **Then** the system raises an error with a message guiding the author to use `{{ conversation }}`
```
Comment (astefanutti) [inline specs/002-structured-events/spec.md]: At that stage, maybe we an afford to not cover that case?

**Review by astefanutti — COMMENTED** [review]
Thanks @rhuss, left few comments, otherwise LGTM!

File: specs/002-structured-events/spec.md (hunk 19 above)
Comment (rhuss) [reply specs/002-structured-events/spec.md:25]: Great idea! Not in scope for this spec PR, but structured events would actually make self-hosting eval easier later since judges could inspect the harness's own tool call patterns. We already have e2e tests that invoke the real Claude API against a fake Jira skill fixture, so this would be the natural next step.

I can work on an e2e test after this PR.

File: specs/002-structured-events/spec.md (hunk 20 above)
Comment (rhuss) [reply specs/002-structured-events/spec.md]: I think `events` is more accurate here. The list contains system init events, tool calls, tool results, and result summaries, not just conversational turns. `turns` implies alternating user/assistant exchanges, but our data model includes non-turn items (system event with model info, result event with cost). It also matches Claude Code's stream-json terminology where each JSONL line is an "event" with a `type` discriminator.

File: specs/002-structured-events/spec.md (hunk 21 above)
Comment (rhuss) [reply specs/002-structured-events/spec.md]: Agreed, let's do a clean cut. `{{ stdout }}` was only introduced in the latest commit and nobody is using it yet. Better to remove it entirely without leaving a migration path for something that never shipped. I'll rip out the `{{ stdout }}` template variable completely from the spec and implementation.

**Comment by rhuss** [comment]
@astefanutti thanks for the feedback, I've addressed those (and removed {{stdout}} completely without migration path as it probably never has been used)

My question now: 

* Should I continue with the implementation on this PR ? (updating then the title to .. [Spec + Impl]) ? This would keep the PR history more focussed and follows the natural flow. 
* Or would you prefer to merge this PR first and then a second (or more) PRs for the implementation ?

I tend to former, we could also leverage labels for that to make the status in the SDD flow more clear than only a title update. wdyt ?

(as said, I'm shamelessly using this project also as a test-bed for a collab SDD workflow, hope this is ok :-))

Do you have any feedback on the `REVIEWERS.md` ? Was it hard to follow ? Did it serve its purpose to help in a spec-only review ? What would you improve or add ? That kind of feedback would be very helpful 🙏

File: specs/002-structured-events/plan.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (rhuss) [reply specs/002-structured-events/plan.md:17]: Out of scope for this spec. The eval harness runs locally on the developer's machine against their own skill outputs, and run artifacts stay in the local `eval/runs/` directory (gitignored). Adding a redaction layer would add complexity without addressing a real threat in our deployment model.

If the harness gains shared artifact storage or CI artifact publishing later, PII redaction would make sense as a separate feature spec.

File: specs/002-structured-events/spec.md (hunk 20 above)
Comment (astefanutti) [reply specs/002-structured-events/spec.md]: I agree, that sounds good. I think that's what also made use `events` initially, but wanted to double check with you before we commit to that name.

File: specs/002-structured-events/spec.md (hunk 19 above)
Comment (astefanutti) [reply specs/002-structured-events/spec.md:25]: Yes, better be something to follow-up.

**Comment by astefanutti** [comment]
> @astefanutti thanks for the feedback, I've addressed those (and removed {{stdout}} completely without migration path as it probably never has been used)
> 
> My question now:
> 
> * Should I continue with the implementation on this PR ? (updating then the title to .. [Spec + Impl]) ? This would keep the PR history more focussed and follows the natural flow.
> * Or would you prefer to merge this PR first and then a second (or more) PRs for the implementation ?
> 
> I tend to former, we could also leverage labels for that to make the status in the SDD flow more clear than only a title update. wdyt ?

Yes, I'd also be inclined for "spec + impl" in the same PR. Maybe to much larger spec / scope the SDD flow could also accommodate for the later.

> (as said, I'm shamelessly using this project also as a test-bed for a collab SDD workflow, hope this is ok :-))

Absolutely :) Love the idea! 

> Do you have any feedback on the `REVIEWERS.md` ? Was it hard to follow ? Did it serve its purpose to help in a spec-only review ? What would you improve or add ? That kind of feedback would be very helpful 🙏

I find it useful and like the Why/What type of articulation. On the what, sometimes it's blurred for relatively small scope how to separate the what from the how. In that example, I think the what might also touch on the how and could potentially be separated. To push the "questions" framing, there could be a When section, maybe for scope/out-of-scope.

**Comment by rhuss** [comment]
thanks that feedback makes fully sense and I updated the REVIEWERS.md generation for the next time. I'm now continuing with the implementation on this PR directly.

**Comment by rhuss** [comment]
@astefanutti Implementation is done and pushed. The PR now includes both spec and implementation (title and labels updated accordingly). Ready for your review when you get a chance.

File: brainstorm/01-structured-events.md [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+
+- Pros: Simplest change, no new file format
+- Cons: Still parses at scoring time, events not inspectable on disk, not reusable outside scoring
+
+### C: Shared Parser Module + Collection-Time Storage
+
+New `agent_eval/events.py` module with a single `parse_stream_events()` function. Called by `collect.py` at collection time, results stored as `events.json`. All scoring consumers use the structured events from the record dict.
+
+- Pros: Single source of truth for JSONL parsing, collection-time storage for inspection, reusable by any future consumer
+- Cons: Larger scope (touches collect.py, score.py, new module), event schema becomes a contract
```
Comment (astefanutti) [inline brainstorm/01-structured-events.md:48]: For "event schema becomes a contract" I wonder how we could / should adopt something like https://github.com/harbor-framework/harbor/blob/main/rfcs/0001-trajectory-format.md eventually.

It seems Claude Code supports it behind an environment variable (`CLAUDE_CODE_ENHANCED_TELEMETRY_BETA`?)

**Review by astefanutti — APPROVED** [review]
Thanks @rhuss, it looks good to me!

I realize it raises the question of adopting emergent trace / trajectory standard format like https://github.com/harbor-framework/harbor/blob/main/rfcs/0001-trajectory-format.md.

That may be useful to start exploring other harnesses like OpenCode to review the options and what could be the best strategy.

I also wonder how we could start bundling those reusable / generic judges.
